app/services/migration/migrators/application.rb

```diff
@@ -30,41 +38,86 @@ def model
+
+        ensure_relationships_are_consistent!(ecf_npq_application, application)
+
+        ecf_schedule = ecf_npq_application.profile&.schedule
+        if ecf_schedule
+          schedule_cohort_id = find_cohort_id!(start_year: ecf_schedule.cohort.start_year)
+          course_group_name = course_groups_by_schedule_type(ecf_schedule.type).name
+          application.schedule_id = find_schedule_id!(cohort_id: schedule_cohort_id, identifier: ecf_schedule.schedule_identifier, course_group_name:)
+        end
+
+        application.cohort_id = find_cohort_id!(start_year: ecf_npq_application.cohort.start_year)
+        application.itt_provider_id = find_itt_provider_id!(legal_name: ecf_npq_application.itt_provider) if ecf_npq_application.itt_provider
+        application.private_childcare_provider_id = find_private_childcare_provider_id!(provider_urn: ecf_npq_application.private_childcare_provider_urn) if ecf_npq_application.private_childcare_provider_urn
 
-        compare_attributes_values!(ecf_npq_application, application)
+        application.training_status = ecf_npq_application.profile&.training_status if ecf_npq_application.profile
+        application.ukprn = ecf_npq_application.school_ukprn
+
+        application.update!(ecf_npq_application.attributes.slice(*ATTRIBUTES))
       end
     end
 
   private
 
-    def applications_by_ecf_id
-      @applications_by_ecf_id ||= ::Application
-        .select(ATTRIBUTES_TO_COMPARE + %i[ecf_id])
-        .where(ecf_id: self.class.ecf_npq_applications.pluck(:id))
-        .index_by(&:ecf_id)
+    def find_schedule_id!(cohort_id:, identifier:, course_group_name:)
+      schedule_ids_by_cohort_id_and_identifier_and_course_group.dig(cohort_id, identifier, course_group_name) || raise(ActiveRecord::RecordNotFound, "Couldn't find Schedule")
     end
 
-    def report_applications_not_in_ecf_as_failures
-      applications_not_in_ecf = ::Application.where.not(ecf_id: self.class.ecf_npq_applications.pluck(:id)).select(:id)
-      applications_not_in_ecf.each { |a| failure_manager.record_failure(a, "NPQApplication not found in ECF") }
+    def schedule_ids_by_cohort_id_and_identifier_and_course_group
+      @schedule_ids_by_cohort_id_and_identifier_and_course_group ||= begin
+        schedules = ::Schedule.includes(:course_group).pluck(:id, :cohort_id, :identifier, "course_groups.name")
+        schedules.each_with_object({}) do |(id, cohort_id, identifier, course_group), hash|
+          hash[cohort_id] ||= {}
+          hash[cohort_id][identifier] ||= {}
+          hash[cohort_id][identifier][course_group] = id
+        end
+      end
     end
 
-    def compare_attributes_values!(ecf_npq_application, application)
-      return unless ATTRIBUTES_TO_COMPARE.any? { |attribute| ecf_npq_application[attribute] != application[attribute] }
+    def ensure_relationships_are_consistent!(ecf_npq_application, application)
+      if application.school_id != find_school_id!(urn: ecf_npq_application.school_urn)
```

Comment:
there are many school failures in the report, I think we can avoid some of them by ignoring `nil` from the lookup. Also worth revisiting why others are failing as schools are there 🤔